File: sndbridge/src/oggopusdecoder.c
#ifdef SNDBRIDGE_OPUS_DECODING
#include <assert.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

#include "oggdecoders.h"


static int read_func(void* stream, unsigned char* ptr, int nbytes) {
    return (int)filehandle_read((FileHandle_t*)stream, ptr, nbytes);
}
static int seek_func(void* stream, opus_int64 offset, int whence) {
    int ret;
    if (filehandle_seek((FileHandle_t*)stream, (_off64_t)offset, whence) < 0)
        ret = -1;
    else
        ret = 0;
    return ret;
}
static opus_int64 tell_func(void* stream) {
    return (opus_int64)filehandle_tell((FileHandle_t*)stream);
}


static const int32_t OPUS_RATE = 48000;
static const double OPUS_RATE_CONV = 1000.0 / 48000;

static struct OpusFileCallbacks callbacks = {
    .read = read_func,
    .seek = seek_func,
    .close = NULL,
    .tell = tell_func};


OggOpusDecoder oggopusdecoder_init(FileHandle_t* file_hnd) {
    OggOpusDecoder oggopusdecoder = malloc(sizeof(OggOpusDecoder_t));
    assert(oggopusdecoder);

    OggOpusFile* op = op_open_callbacks((void*)file_hnd, &callbacks, NULL, 0, NULL);
    if (!op) {
        free(oggopusdecoder);
        return NULL;
    }

    oggopusdecoder->op = op;
    oggopusdecoder->duration = op_pcm_total(oggopusdecoder->op, -1) * OPUS_RATE_CONV;
    oggopusdecoder->channels = op_channel_count(oggopusdecoder->op, -1);
    oggopusdecoder->file_hnd = file_hnd;
    oggopusdecoder->bytes_per_channel = op_channel_count(oggopusdecoder->op, -1) * 2;

    return oggopusdecoder;
}

void oggopusdecoder_destroy(OggOpusDecoder oggopusdecoder) {
    op_free(oggopusdecoder->op);
    free(oggopusdecoder);
}


int32_t oggopusdecoder_read(OggOpusDecoder oggopusdecoder, float* buffer, int32_t samples_per_channel) {
    int buf_size = samples_per_channel * oggopusdecoder->channels;
    return op_read_float(oggopusdecoder->op, buffer, buf_size, NULL);
}

void oggopusdecoder_get_info(OggOpusDecoder oggopusdecoder, int32_t* rate, int32_t* channels, double* duration) {
    *rate = OPUS_RATE;
    *channels = oggopusdecoder->channels;
    *duration = oggopusdecoder->duration;
}

bool oggopusdecoder_seek(OggOpusDecoder oggopusdecoder, double timestamp) {
    ogg_int64_t pcm_offset = (ogg_int64_t)(timestamp / OPUS_RATE_CONV);
    return op_pcm_seek(oggopusdecoder->op, pcm_offset);
}
#endif

File: src-desktop/soundbridge/oggopusdecoder.cs
using System;
using System.IO;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using Engine.Externals.SoundBridge.Interop;

namespace Engine.Externals.SoundBridge;

internal unsafe class OGGOpusDecoder : IDecoder {
    private const int OPUS_RATE = 48000;

    private OggOpusFile* op;
    private uint bytes_per_channel, channels;
    private double duration;
    private long loop_start, loop_length;
    private nint src_hnd;

    private OGGOpusDecoder() { }

    public static OGGOpusDecoder Init(ISourceHandle src_hnd) {
        OGGOpusDecoder oggopusdecoder = new OGGOpusDecoder();

        OpusFileCallbacks callbacks = new OpusFileCallbacks() {
            read = &read_func,
            seek = &seek_func,
            tell = &tell_func,
            close = null
        };

        GCHandle gchandle = GCHandle.Alloc(src_hnd, GCHandleType.Normal);
        oggopusdecoder.src_hnd = GCHandle.ToIntPtr(gchandle);

        OggOpusFile* op = Opusfile.op_open_callbacks(oggopusdecoder.src_hnd, &callbacks, null, 0, null);
        if (op == null) {
            gchandle.Free();
            return null;
        }

        oggopusdecoder.op = op;
        oggopusdecoder.duration = Opusfile.op_pcm_total(oggopusdecoder.op, -1) * OGGOpusDecoder.OPUS_RATE;
        oggopusdecoder.channels = (uint)Opusfile.op_channel_count(oggopusdecoder.op, -1);
        oggopusdecoder.bytes_per_channel = oggopusdecoder.channels * 2;
        oggopusdecoder.loop_start = 0;
        oggopusdecoder.loop_length = 0;

        // find loop points
        OpusTags* comments_info = Opusfile.op_tags(oggopusdecoder.op, -1);
        for (int i = 0 ; i < comments_info->comments ; i++) {
            int length = comments_info->comment_lengths[i];
            string comment = Marshal.PtrToStringUTF8((nint)comments_info->user_comments[i], length);

            if (length < 1 || comment == null) continue;

            if (comment.StartsWith(OggUtil.LOOPSTART)) {
                Int64.TryParse(comment.AsSpan(OggUtil.LOOPSTART.Length), out oggopusdecoder.loop_start);
            }

            if (comment.StartsWith(OggUtil.LOOPLENGTH)) {
                Int64.TryParse(comment.AsSpan(OggUtil.LOOPLENGTH.Length), out oggopusdecoder.loop_length);
            }
        }

        return oggopusdecoder;
    }

    public void Dispose() {
        Opusfile.op_free(this.op);
        GCHandle.FromIntPtr(this.src_hnd).Free();
    }


    public int Read(float* buffer, uint samples_per_channel) {
        uint buf_size = samples_per_channel * this.channels;
        return Opusfile.op_read_float(this.op, buffer, (int)buf_size, null);
    }

    public SampleFormat GetInfo(out uint rate, out uint channels, out double duration) {
        rate = OGGOpusDecoder.OPUS_RATE;
        channels = this.channels;
        duration = this.duration;
        return SampleFormat.FLOAT32;
    }

    public bool Seek(double seconds) {
        long pcm_offset = (long)(seconds / OGGOpusDecoder.OPUS_RATE);
        return Opusfile.op_pcm_seek(this.op, pcm_offset) != 0;
    }

    public void GetLoopPoints(out long loop_start, out long loop_length) {
        loop_start = this.loop_start;
        loop_length = this.loop_length;
    }


    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private static ISourceHandle recover(nint gchandle_ptr) {
        return (ISourceHandle)GCHandle.FromIntPtr(gchandle_ptr).Target;
    }


    [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvCdecl) })]
    private static int read_func(nint stream, byte* ptr, int nbytes) {
        ISourceHandle src_hnd = recover(stream);
        return src_hnd.Read(ptr, nbytes);
    }

    [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvCdecl) })]
    private static int seek_func(nint stream, long offset, int whence) {
        ISourceHandle src_hnd = recover(stream);
        return src_hnd.Seek(offset, (SeekOrigin)whence);
    }

    [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvCdecl) })]
    private static long tell_func(nint stream) {
        ISourceHandle src_hnd = recover(stream);
        return src_hnd.Tell();
    }

}

